### PR TITLE
Improve error for else (cond) {}

### PR DESCRIPTION
**Pull request by singalhimanshu**
Fix: #578

### CODE REVIEW

File: src/quick-lint-js/error.h [hunk 1]
```diff
@@ -926,6 +926,10 @@
       WARNING(QLJS_TRANSLATABLE("redundant delete statement on variable"),     \
               delete_expression))                                              \
                                                                                \
+  QLJS_ERROR_TYPE(                                                             \
+      error_missing_if_after_else, "E0184", { source_code_span expected_if; }, \
+      ERROR(QLJS_TRANSLATABLE("missing if after else"), expected_if))          \
```
Comment (strager) [inline src/quick-lint-js/error.h:931]: Write quotes around code:
```suggestion
      ERROR(QLJS_TRANSLATABLE("missing 'if' after 'else'"), expected_if))          \
```

File: test/test-parse-statement.cpp [hunk 2]
```diff
@@ -715,6 +715,25 @@ TEST(test_parse, else_without_if) {
   }
 }
 
+TEST(test_parse, missing_if_after_else) {
+  {
+    spy_visitor v;
+    padded_string code(u8"if (false) {} else (true) {}"_sv);
```
Comment (strager) [inline test/test-parse-statement.cpp:721]: Nit: add test cases:

* should *not* report for `if (false) {} else true {}` (it should complain about `{` instead)
* should *not* report anything for `if (false) {} else (true)NEWLINE{}`

File: test/test-parse-statement.cpp [hunk 3]
```diff
@@ -715,6 +715,25 @@ TEST(test_parse, else_without_if) {
...
+    parser p(&code, &v);
+    EXPECT_TRUE(p.parse_and_visit_statement(v));
+    EXPECT_THAT(v.visits, ElementsAre("visit_enter_block_scope",  //
+                                      "visit_exit_block_scope"));
+    EXPECT_THAT(
+        v.errors,
+        ElementsAre(ERROR_TYPE_OFFSETS(
+                        &code, error_missing_semicolon_after_statement,  //
+                        where, strlen(u8"if (false) {} else (true)"), u8""),
```
Comment (strager) [inline test/test-parse-statement.cpp:730]: Must fix: The point of error_missing_if_after_else is to replace error_missing_semicolon_after_statement. We should report one or the other, not both.

File: src/quick-lint-js/parse.h [hunk 4]
```diff
@@ -2808,7 +2808,16 @@ class parser {
 
     if (this->peek().type == token_type::kw_else) {
       this->skip();
+      const char8 *end_of_else = this->lexer_.end_of_previous_token();
+      bool has_left_paren = this->peek().type == token_type::left_paren;
       parse_and_visit_body();
+      bool has_left_curly = this->peek().type == token_type::left_curly;
+      if (this->peek().has_leading_newline == false && has_left_paren &&
+          has_left_curly) {
+        this->error_reporter_->report(error_missing_if_after_else{
```
Comment (strager) [inline src/quick-lint-js/parse.h:2817]: Nit: Include a comment with an example code snippet.
```suggestion
        // if (cond) {} else (cond) {}  // Invalid.
        this->error_reporter_->report(error_missing_if_after_else{
```

File: src/quick-lint-js/parse.h [hunk 5]
```diff
@@ -2808,7 +2808,16 @@ class parser {
 
     if (this->peek().type == token_type::kw_else) {
       this->skip();
+      const char8 *end_of_else = this->lexer_.end_of_previous_token();
+      bool has_left_paren = this->peek().type == token_type::left_paren;
       parse_and_visit_body();
```
Comment (strager) [inline src/quick-lint-js/parse.h:2813]: Idea to fix error_missing_semicolon_after_statement:
```suggestion
      if (has_left_paren) {
        // The body of the 'else' is an expression.
        // copy-pasta from https://github.com/quick-lint/quick-lint-js/blob/7d32d9806fc70cc29bcaf5eaea921302860f9418/src/quick-lint-js/parse.h#L394-L395
        this->parse_and_visit_expression(v);
        // Don't do this: parse_expression_end();  // calls this->consume_semicolon();, which is what's reporting the extraneous error
      } else {
        parse_and_visit_body();
      }
```

**Review by strager — CHANGES_REQUESTED** [review]

File: src/quick-lint-js/parse.h [hunk 6]
```diff
@@ -2810,10 +2810,15 @@ class parser {
...
+      } else {
+        parse_and_visit_body();
+      }
       bool has_left_curly = this->peek().type == token_type::left_curly;
       if (this->peek().has_leading_newline == false && has_left_paren &&
           has_left_curly) {
+        // if (cond) {} else (cond) {} // Invalid
         this->error_reporter_->report(error_missing_if_after_else{
             .expected_if = source_code_span(end_of_else, end_of_else),
         });
```
Comment (strager) [inline src/quick-lint-js/parse.h:2824]: Nit: I think this would be easier to understand if it was moved up to after line 2814. What do you think?

**Review by strager — APPROVED** [review]
Looks good to me! Let me know when I should merge.

File: src/quick-lint-js/parse.h (hunk 6 above)
Comment (singalhimanshu) [reply src/quick-lint-js/parse.h:2824]: I think it makes sense here.

**Comment by singalhimanshu** [comment]
> Looks good to me! Let me know when I should merge.

Ship it :rocket:

**Pull request merged**